dimension_label.go

```diff
@@ -0,0 +1,405 @@
+	}
+
+	freeOnGC(&dimLabel)
+	return &dimLabel, nil
+}
+
+// DimensionLabelFromName retrieves a dimension label from an array schema with the requested name.
+func (a *ArraySchema) DimensionLabelFromName(name string) (*DimensionLabel, error) {
+	cAttrName := C.CString(name)
+	defer C.free(unsafe.Pointer(cAttrName))
+	dimLabel := DimensionLabel{context: a.context}
+	ret := C.tiledb_array_schema_get_dimension_label_from_name(a.context.tiledbContext, a.tiledbArraySchema,
+		cAttrName, &dimLabel.tiledbDimensionLabel)
+	if ret != C.TILEDB_OK {
+		return nil, fmt.Errorf("Error getting dimension label '%s' for ArraySchema: %s", name, a.context.LastError())
+	}
+
+	freeOnGC(&dimLabel)
+	return &dimLabel, nil
+}
+
+// HasDimensionLabel checks whether the array schema has a dimension label of the given name.
+func (a *ArraySchema) HasDimensionLabel(name string) (bool, error) {
+	cLabelName := C.CString(name)
+	defer C.free(unsafe.Pointer(cLabelName))
+
+	var hasLabel C.int32_t
+	ret := C.tiledb_array_schema_has_dimension_label(a.context.tiledbContext, a.tiledbArraySchema,
+		cLabelName, &hasLabel)
+	if ret != C.TILEDB_OK {
+		return false, fmt.Errorf("Error checking ArraySchema for dimension label '%s': %s", name, a.context.LastError())
+	}
+
+	return hasLabel != 0, nil
+}
+
+// DimensionLabelsNum returns the number of dimension label in this array schema
+func (a *ArraySchema) DimensionLabelsNum() (uint64, error) {
+	var labelNum uint64
+
+	ret := C.tiledb_array_schema_get_dimension_label_num(a.context.tiledbContext, a.tiledbArraySchema, (*C.uint64_t)(unsafe.Pointer(&labelNum)))
+	if ret != C.TILEDB_OK {
+		return 0, fmt.Errorf("Error fetching dimension label number: %s", a.context.LastError())
+	}
+
+	return labelNum, nil
+}
+
+// SetDimensionLabelFilterList sets a filter on a dimension label filter in an array schema.
+func (a *ArraySchema) SetDimensionLabelFilterList(name string, filterList FilterList) error {
+	cName := C.CString(name)
+	defer C.free(unsafe.Pointer(cName))
+
+	ret := C.tiledb_array_schema_set_dimension_label_filter_list(a.context.tiledbContext, a.tiledbArraySchema,
+		cName, filterList.tiledbFilterList)
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error setting dimension label filter list on ArraySchema: %s", a.context.LastError())
+	}
+
+	return nil
+}
+
+// SetDimensionLabelTileExtent sets the tile extent for the dimension label
+func (a *ArraySchema) SetDimensionLabelTileExtent(labelName string, dimType Datatype, extent interface{}) error {
+	cName := C.CString(labelName)
+	defer C.free(unsafe.Pointer(cName))
+
+	extentType := reflect.TypeOf(extent).Kind()
+	if extentType != dimType.ReflectKind() {
+		return fmt.Errorf("Dimension and extent do not have the same data types. Dimension: %s, Extent: %s",
+			dimType.ReflectKind(), extentType)
+	}
+
+	// Create extent void*
+	var cExtent unsafe.Pointer
+	switch dimType {
+	case TILEDB_INT8:
+		tmpExtent := extent.(int8)
```

Comment:
There are a few things going on here which will be annoying to deal with:

- Since Go is garbage-collected, memory becomes eligible for freeing (and thus should be considered as already freed) after the last time they are referred to.
- `unsafe.Pointer` intentionally is opaque to the GC, so something referred to by an `unsafe.Pointer` does not count as actually there.

This means that in this case:

 1. `tmpExtent` gets space allocated (on the stack, in this case), and the value of `extent` is copied into it.
 2. We create an `unsafe.Pointer` to `tmpExtent`.
 3. `tmpExtent` is no longer referred to, thus making it eligible for collection.
 4. We pass that pointer out to TileDB.

(This might be something we do elsewhere in this package; if so we need to fix it there too.)

In this case we need to ensure `tmpExtent` stays around. So I think we need to do something like:

```go
var extentPtr any
defer runtime.KeepAlive(extentPtr)

// Here we can also reduce the amount of repetition by using a type-switch.
switch extent := extent.(type) {
case int8:
  // This ensures that `extent` stays alive.
  extentPtr = &extent
  // and since cExtent is a pointer to the same thing as extentPtr, it's OK.
  cExtent := unsafe.Pointer(&extent)
// ...
}
```

(This is annoying and fiddly.)

Reply:
Done, thanks! Similar pattern was also in dimension.go, I made this change there too.